packages/core/src/codewhisperer/util/editorContext.ts

```diff
@@ -23,19 +23,83 @@ import { AuthUtil } from './authUtil'
 
 let tabSize: number = getTabSizeSetting()
 
+const languageCommentChars: Record<string, string> = {
+    python: '# ',
+    java: '// ',
+}
+
+export function extractSingleCellContext(cell: vscode.NotebookCell, referenceLanguage?: string): string {
+    // Extract the text verbatim if the cell is code and the cell has the same language.
+    // Otherwise, add the correct comment string for the refeference language
+    const cellText = cell.document.getText()
+    if (
+        cell.kind === vscode.NotebookCellKind.Markup ||
+        (runtimeLanguageContext.normalizeLanguage(cell.document.languageId) ?? cell.document.languageId) !==
+            referenceLanguage
+    ) {
+        const commentPrefix = (referenceLanguage && languageCommentChars[referenceLanguage]) ?? ''
+        if (commentPrefix === '') {
+            return cellText
+        }
+        return cell.document
+            .getText()
+            .split('\n')
+            .map((line) => `${commentPrefix}${line}`)
+            .join('\n')
+    }
+    return cellText
+}
+
+export function extractCellsSliceContext(
```

Comment:
"notebook" is a useful nonce that helps discovery. especially since these functions are mingled with non-notebook logic.

```suggestion
export function getNotebookCellsSliceContext(
```